src/proxy/server.rs

```diff
@@ -190,164 +188,155 @@ impl fmt::Display for NoOriginalDst {
             local: connection.local_addr().unwrap_or(self.listen_addr),
-            orig_dst,
+            orig_dst: connection.original_dst_addr(),
             tls_peer: connection.peer_identity(),
             _p: (),
         };
 
+        let should_detect = connection.should_detect_protocol();
         let io = self.accept.accept(&source, connection);
 
-        let connect = self.connect.clone();
-
-        if disable_protocol_detection {
-            trace!("protocol detection disabled for {:?}", orig_dst);
-            let fwd = tcp::forward(io, connect, source);
-            let fut = self.drain_signal.clone().watch(fwd, |_| {});
-            return log.future(Either::B(fut));
-        }
+        let accept_fut = if should_detect {
+            Either::A(
+                io.peek()
+                    .map_err(|e| debug!("peek error: {}", e))
+                    .map(|io| {
+                        let proto = Protocol::detect(io.peeked());
+                        (proto, io)
+                    }),
+            )
+        } else {
+            debug!("protocol detection disabled for {:?}", source.orig_dst);
+            Either::B(future::ok((None, io)))
+        };
 
-        let detect_protocol = io
-            .peek()
-            .map_err(|e| debug!("peek error: {}", e))
-            .map(|io| {
-                let p = Protocol::detect(io.peeked());
-                (p, io)
-            });
+        let connect = self.connect.clone();
+        let log = self.log.clone().with_remote(source.remote);
 
         let mut http = self.http.clone();
-        let mut route = self.route.clone();
-        let drain_signal = self.drain_signal.clone();
+        let mut make_http = self.make_http.clone();
         let log_clone = log.clone();
-        let serve = detect_protocol.and_then(move |(proto, io)| match proto {
-            None => Either::A({
+        let initial_stream_window_size = self.h2_settings.initial_stream_window_size;
+        let initial_conn_window_size = self.h2_settings.initial_connection_window_size;
+        let serve_fut = accept_fut.and_then(move |(proto, io)| match proto {
+            None => {
                 trace!("did not detect protocol; forwarding TCP");
                 let fwd = tcp::forward(io, connect, source);
-                drain_signal.watch(fwd, |_| {})
-            }),
+                Either::A(drain.watch(fwd, |_| {}))
+            }
 
-            Some(proto) => Either::B(match proto {
-                Protocol::Http1 => Either::A({
-                    trace!("detected HTTP/1");
-                    route
-                        .make_service(source)
-                        .map_err(|never| match never {})
-                        .and_then(move |s| {
+            Some(proto) => Either::B({
+                make_http
+                    .make_service(source)
+                    .map_err(|never| match never {})
+                    .and_then(move |http_svc| match proto {
```

Comment:
Nit/TIOLI: It would be nice if this was not nested so many indent levels deep...maybe this closure could factored out into a higher level `let` binding or something?